modules/oracle/oracle.go

```diff
@@ -0,0 +1,170 @@
+
+	"github.com/chebyrash/promise"
+)
+
+const (
+	// 10 minutes = 600 seconds, 3s for every new block.
+	// A tick is broadcasted every 200 blocks produced.
+	priceOracleBroadcastInterval = uint64(600 / 3)
+	priceOraclePollInterval      = time.Second * 15
+
+	// 10 minutes = 600 seconds, 3s for every new block
+	btcChainRelayInterval = uint64(600 / 3)
+)
+
+var (
+	watchSymbols = []string{"BTC", "ETH", "LTC"}
+)
+
+type BlockSync interface {
+	RegisterBlockTick(string, vstream.BTFunc, bool)
+}
+
+type BlockSchedule interface {
+	GetSchedule(slotHeight uint64) []stateEngine.WitnessSlot
+}
+
+type Oracle struct {
+	p2pServer  *libp2p.P2PServer
+	oracleP2P  p2p.OracleP2pParams
+	service    libp2p.PubSubService[p2p.Msg]
+	conf       common.IdentityConfig
+	electionDb elections.Elections
+	witness    witnesses.Witnesses
+
+	vStream     BlockSync
+	stateEngine BlockSchedule
+
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+
+	priceOracle          *price.PriceOracle
+	broadcastPriceSignal chan blockTickSignal
+
+	// to be used within a node, for price querying from API endpoints
+	observePriceChan chan []p2p.ObservePricePoint
+
+	// to be used within the network, for broadcasting average prices
+	broadcastPriceChan chan []p2p.AveragePricePoint
+
+	// for block signatures
+	priceBlockSignatureChan chan p2p.VSCBlock
+
+	// for new block
+	broadcastPriceBlockChan chan p2p.VSCBlock
+
+	// for communication between nodes in a network
+	msgChan chan p2p.Msg
+}
+
+func New(
+	p2pServer *libp2p.P2PServer,
+	oracleP2P p2p.OracleP2pParams,
+	conf common.IdentityConfig,
+	electionDb elections.Elections,
+	witness witnesses.Witnesses,
+	vstream BlockSync,
+	stateEngine BlockSchedule,
+) *Oracle {
+	return &Oracle{
+		p2pServer:   p2pServer,
+		oracleP2P:   oracleP2P,
+		conf:        conf,
+		electionDb:  electionDb,
+		witness:     witness,
+		vStream:     vstream,
+		stateEngine: stateEngine,
+
+		msgChan: make(chan p2p.Msg, 1),
+
+		observePriceChan:        make(chan []p2p.ObservePricePoint, 128),
+		broadcastPriceChan:      make(chan []p2p.AveragePricePoint, 128),
+		broadcastPriceSignal:    make(chan blockTickSignal, 1),
+		priceBlockSignatureChan: make(chan p2p.VSCBlock, 1024),
+		broadcastPriceBlockChan: make(chan p2p.VSCBlock, 1024),
+	}
+}
+
+// Init implements aggregate.Plugin.
+// Runs initialization in order of how they are passed in to `Aggregate`
+func (o *Oracle) Init() error {
+	const userCurrency = "usd" // NOTE: only supporting USD for now
+
+	var err error
+
+	o.priceOracle, err = price.New(userCurrency)
+	if err != nil {
+		return fmt.Errorf("failed to initialize price oracle: %w", err)
+	}
+
+	// o.blockRelay = btcrelay.New(o.blockRelayChan)
+
+	o.ctx, o.cancelFunc = context.WithCancel(context.Background())
+
+	return nil
+}
+
+// Start implements aggregate.Plugin.
+// Runs startup and should be non blocking
+func (o *Oracle) Start() *promise.Promise[any] {
+	o.vStream.RegisterBlockTick("oracle", o.blockTick, false)
```

Comment:
fyi RegisterBlockTick should be async for this type of operation. It's not mission critical thus should run asynchronously